# Návštěvnost turistických cílů v Olomouckém kraji

In [36]:
import pandas

Načteme si data, která udávají návštěvnost různých turistických cílů v Olomouckém kraji. Zde je  odkaz na zdroj dat: https://data.gov.cz/datov%C3%A1-sada?iri=https%3A%2F%2Fdata.gov.cz%2Fzdroj%2Fdatov%C3%A9-sady%2F60609460%2F671456014 (tlačítko Stáhnout). 

Všimni si, že data jsou oddělená středníky, a také jsou uložená v jiném kódování, než jsme zvyklí, proto musíme toto kódování specifikovat, aby byla čeština čitelná s veškerou diakritikou:

In [37]:
df = pandas.read_csv("navstevnost-tur-cilu.csv", sep=";", encoding="cp1250")
df

Unnamed: 0                           Objekt     2018     2017     2016  \
0       Hrady                      Hrad Bouzov  111 668  111 601  121 532   
1         NaN                Hrad Helf�t�n   77 378   86 341   93 244   
2         NaN                 Hrad �ternberk   56 712   47 095   60 689   
3     Z�mky  Z�mek �echy pod Kos��em   44 278   46 966   33 573   
4         NaN       Z�mek N�m욝 na Han�   27 123   28 928   28 855   
..        ...                              ...      ...      ...      ...   
94        NaN   Meziro�n� srovn�n� (%)     100%      98%     106%   
95        NaN                              NaN      NaN      NaN      NaN   
96        NaN     x - subjekt �daj neposkytl      NaN      NaN      NaN   
97        NaN                              NaN      NaN      NaN      NaN   
98        NaN                              NaN      NaN      NaN      NaN   

       2015     2014    2013     2012    2011  ...     2009     2008     2007  \
0   105 519  104 870  91 341  100 743  93 589  ...  107 415  100 401  110 904   
1    89 253   90 521  75 762   79 440  90 524  ...   94 687  100 722  107 702   
2    55 078   53 652  46 327   47 101  45 854  ...   57 984   38 076   36 674   
3       NaN      NaN     NaN      NaN   4 918  ...    5 857    5 091    5 541   
4    26 004   24 212  25 018   27 097  25 672  ...   29 102   29 109   29 215   
..      ...      ...     ...      ...     ...  ...      ...      ...      ...   
94     112%     110%     99%      94%    108%  ...     106%     100%     107%   
95      NaN      NaN     NaN      NaN     NaN  ...      NaN      NaN      NaN   
96      NaN      NaN     NaN      NaN     NaN  ...      NaN      NaN      NaN   
97      NaN      NaN     NaN      NaN     NaN  ...      NaN      NaN      NaN   
98      NaN      NaN     NaN      NaN     NaN  ...      NaN      NaN      NaN   

       2006     2005    2004     2003     2002     2001     2000  
0   110 493  102 260  83 540  102 807  118 101  175 112  129 830  
1    94 203   96 357  95 500  102 923   99 385   89 607   99 963  
2    36 310   20 583  16 282   16 316   17 806   17 754   20 341  
3     5 169      NaN     NaN      NaN      NaN      NaN      NaN  
4    24 063   26 067  22 605   20 695   20 120   17 500   16 500  
..      ...      ...     ...      ...      ...      ...      ...  
94     100%     128%     95%      93%     101%     103%      NaN  
95      NaN      NaN     NaN      NaN      NaN      NaN      NaN  
96      NaN      NaN     NaN      NaN      NaN      NaN      NaN  
97      NaN      NaN     NaN      NaN      NaN      NaN      NaN  
98      NaN      NaN     NaN      NaN      NaN      NaN      NaN  

[99 rows x 21 columns]

Podle prvního sloupce to vypadá, že se jedná o víc tabulek spojených dohromady. Typ objektu je udaný vždy jen v první řádce, a pak následují prázdné hodnoty až do sekce s dalším objektem. Nám by se hodilo si data doplnit, abychom mohli jednoduše filtrovat podle typu objektu.

### Cvičení

* V první řadě přejmenuj první sloupec na něco vypovídajícího, třeba `"Typ objektu"`.
* Dále vyplň chybějící hodnoty: Aplikuj metodu `fillna` na sloupec `"Typ objektu"` v kombinaci s parametrem `method="ffill"`. Podívej se na dokumentaci metody `fillna` pro příklad použití: https://pandas.pydata.org/docs/reference/api/pandas.DataFrame.fillna.html

In [38]:
df = df.rename({"Unnamed: 0": "Typ objektu"}, axis="columns")
df.columns

Index(['Typ objektu', 'Objekt', '2018', '2017', '2016', '2015', '2014', '2013',
       '2012', '2011', '2010', '2009', '2008', '2007', '2006', '2005', '2004',
       '2003', '2002', '2001', '2000'],
      dtype='object')

In [39]:
df["Typ objektu"] = df["Typ objektu"].fillna(method="ffill")
df

Typ objektu                           Objekt     2018     2017     2016  \
0        Hrady                      Hrad Bouzov  111 668  111 601  121 532   
1        Hrady                Hrad Helf�t�n   77 378   86 341   93 244   
2        Hrady                 Hrad �ternberk   56 712   47 095   60 689   
3      Z�mky  Z�mek �echy pod Kos��em   44 278   46 966   33 573   
4      Z�mky       Z�mek N�m욝 na Han�   27 123   28 928   28 855   
..         ...                              ...      ...      ...      ...   
94   ostatn�   Meziro�n� srovn�n� (%)     100%      98%     106%   
95   ostatnďż˝                              NaN      NaN      NaN      NaN   
96   ostatn�     x - subjekt �daj neposkytl      NaN      NaN      NaN   
97   ostatnďż˝                              NaN      NaN      NaN      NaN   
98   ostatnďż˝                              NaN      NaN      NaN      NaN   

       2015     2014    2013     2012    2011  ...     2009     2008     2007  \
0   105 519  104 870  91 341  100 743  93 589  ...  107 415  100 401  110 904   
1    89 253   90 521  75 762   79 440  90 524  ...   94 687  100 722  107 702   
2    55 078   53 652  46 327   47 101  45 854  ...   57 984   38 076   36 674   
3       NaN      NaN     NaN      NaN   4 918  ...    5 857    5 091    5 541   
4    26 004   24 212  25 018   27 097  25 672  ...   29 102   29 109   29 215   
..      ...      ...     ...      ...     ...  ...      ...      ...      ...   
94     112%     110%     99%      94%    108%  ...     106%     100%     107%   
95      NaN      NaN     NaN      NaN     NaN  ...      NaN      NaN      NaN   
96      NaN      NaN     NaN      NaN     NaN  ...      NaN      NaN      NaN   
97      NaN      NaN     NaN      NaN     NaN  ...      NaN      NaN      NaN   
98      NaN      NaN     NaN      NaN     NaN  ...      NaN      NaN      NaN   

       2006     2005    2004     2003     2002     2001     2000  
0   110 493  102 260  83 540  102 807  118 101  175 112  129 830  
1    94 203   96 357  95 500  102 923   99 385   89 607   99 963  
2    36 310   20 583  16 282   16 316   17 806   17 754   20 341  
3     5 169      NaN     NaN      NaN      NaN      NaN      NaN  
4    24 063   26 067  22 605   20 695   20 120   17 500   16 500  
..      ...      ...     ...      ...      ...      ...      ...  
94     100%     128%     95%      93%     101%     103%      NaN  
95      NaN      NaN     NaN      NaN      NaN      NaN      NaN  
96      NaN      NaN     NaN      NaN      NaN      NaN      NaN  
97      NaN      NaN     NaN      NaN      NaN      NaN      NaN  
98      NaN      NaN     NaN      NaN      NaN      NaN      NaN  

[99 rows x 21 columns]

Teď můžeme jednoduše pomocí dotazu vyfiltrovat všechny objekty jednoho typu. Můžeš si vybrat jaký chceš, třeba zámky:

In [40]:
zamky = df[df["Typ objektu"]=="Z�mky"]
zamky

Typ objektu                           Objekt    2018    2017    2016  \
3      Z�mky  Z�mek �echy pod Kos��em  44 278  46 966  33 573   
4      Z�mky       Z�mek N�m욝 na Han�  27 123  28 928  28 855   
5      Z�mky                  Z�mek Plumlov  25 853  25 584  19 897   
6      Z�mky                Z�mek Tova�ov  22 954  21 948  23 030   
7      Z�mky                   Z�mek Konice   5 276     NaN     NaN   
8      Z�mky                   Z�mek �sov  41 228  39 546  33 864   
9      Z�mky          Z�mek J�nsk� Vrch  26 598  30 034  24 673   
10     Z�mky           Z�mek Velk� Losiny  35 476  34 896  33 271   

      2015    2014    2013    2012    2011  ...    2009    2008    2007  \
3      NaN     NaN     NaN     NaN   4 918  ...   5 857   5 091   5 541   
4   26 004  24 212  25 018  27 097  25 672  ...  29 102  29 109  29 215   
5   15 300  16 056  15 260  11 863  10 000  ...   5 600  20 000  11 000   
6   21 353  15 149  21 073  23 959  29 697  ...  27 365  28 860  20 170   
7      NaN     NaN     NaN     NaN     NaN  ...     NaN     NaN     NaN   
8   31 957  36 960  20 603  19 784  24 680  ...  43 781  36 256  35 165   
9   23 703  27 515  26 480  28 022  32 329  ...  27 109  33 312  32 955   
10  32 821  31 610  28 407  25 806  29 363  ...  30 645  31 072  35 639   

      2006    2005    2004    2003    2002      2001    2000  
3    5 169     NaN     NaN     NaN     NaN       NaN     NaN  
4   24 063  26 067  22 605  20 695  20 120    17 500  16 500  
5    5 500   5 990     NaN     NaN     NaN       NaN     NaN  
6   20 750  23 187  16 882  21 500  23 000  18�000  16 500  
7      NaN     NaN     NaN     NaN     NaN       NaN     NaN  
8   32 907  91 135  45 167     NaN  41 966    42 918  46 250  
9   33 953  31 012  31 177  32 940  37 081    40 771  35 960  
10  35 841  39 124  41 982  42 813  43 735    46 213  46 715  

[8 rows x 21 columns]

In [41]:
zamky.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 8 entries, 3 to 10
Data columns (total 21 columns):
 #   Column       Non-Null Count  Dtype 
---  ------       --------------  ----- 
 0   Typ objektu  8 non-null      object
 1   Objekt       8 non-null      object
 2   2018         8 non-null      object
 3   2017         7 non-null      object
 4   2016         7 non-null      object
 5   2015         6 non-null      object
 6   2014         6 non-null      object
 7   2013         6 non-null      object
 8   2012         6 non-null      object
 9   2011         7 non-null      object
 10  2010         7 non-null      object
 11  2009         7 non-null      object
 12  2008         7 non-null      object
 13  2007         7 non-null      object
 14  2006         7 non-null      object
 15  2005         6 non-null      object
 16  2004         5 non-null      object
 17  2003         4 non-null      object
 18  2002         5 non-null      object
 19  2001         5 non-null      obj

Trochu matoucí může být, že návštěvnost je udaná jako typ řetězce, místo číselného typu. Představ si to jako kdybychom měli hodnotu načtenou od uživatele `"28 407"`. Jednalo by se o řetězec, který bychom chtěli převést na číselný typ.

### Cvičení

* Napiš funkci `convert_object_series_to_float`, která vezme jeden parametr, tím bude série (v našem případě sloupec). Provede následující operace a upravený řádek vrátí:
  * `str.split().str.join('')` - takto se zbavíme různých mezer
  * `astype(float)` - takto převedeme řetězce v sérii na čísla. Schválně používáme typ `float`, který nám umožní zachovat chybějící hodnoty `NaN`. 

In [48]:
def convert_object_series_to_float(series: pandas.Series) -> pandas.Series:
    series = series.str.split().str.join('')
    series = series.astype(float)
    return series

Vyzkoušej svojí funkci - vytvoříme si testovací sérii, která obsahuje řetězce (typ je tedy `object`):

In [43]:
test_series = pandas.Series(["1 000", "12 345", "1.0"])
test_series

0     1 000
1    12 345
2       1.0
dtype: object

Po zavolání by nám funkce měla vrátit sérii se stejnou délkou a stejným obsahem, akorát tentokrát se jedná o číselné hodnoty - typ `float64`:

In [49]:
convert_object_series_to_float(test_series)

0     1000.0
1    12345.0
2        1.0
dtype: float64

Pojďme využít funkci na návštěvnost zámků.

### Cvičení

* Zbav se sloupce `Typ objektu`. Pro naší tabulku zámků je nevypovídající.
* Nastav sloupec `Objekt` jako index. Tím zajistíme, že obsah tabulky jsou jen data o návštěvnosti.

In [45]:
zamky = zamky.drop("Typ objektu", axis="columns")
zamky.head()

Objekt    2018    2017    2016    2015    2014  \
3  Z�mek �echy pod Kos��em  44 278  46 966  33 573     NaN     NaN   
4       Z�mek N�m욝 na Han�  27 123  28 928  28 855  26 004  24 212   
5                  Z�mek Plumlov  25 853  25 584  19 897  15 300  16 056   
6                Z�mek Tova�ov  22 954  21 948  23 030  21 353  15 149   
7                   Z�mek Konice   5 276     NaN     NaN     NaN     NaN   

     2013    2012    2011    2010    2009    2008    2007    2006    2005  \
3     NaN     NaN   4 918   4 665   5 857   5 091   5 541   5 169     NaN   
4  25 018  27 097  25 672  29 365  29 102  29 109  29 215  24 063  26 067   
5  15 260  11 863  10 000  10 000   5 600  20 000  11 000   5 500   5 990   
6  21 073  23 959  29 697  17 188  27 365  28 860  20 170  20 750  23 187   
7     NaN     NaN     NaN     NaN     NaN     NaN     NaN     NaN     NaN   

     2004    2003    2002      2001    2000  
3     NaN     NaN     NaN       NaN     NaN  
4  22 605  20 695  20 120    17 500  16 500  
5     NaN     NaN     NaN       NaN     NaN  
6  16 882  21 500  23 000  18�000  16 500  
7     NaN     NaN     NaN       NaN     NaN

In [50]:
zamky = zamky.apply(convert_object_series_to_float)
zamky.info()

ValueError: could not convert string to float: 'Z�mek�echypodKos��em'

Teď můžeme bezpečně aplikovat naší funkci `convert_object_series_to_float` na všechny sloupce:

In [ ]:
zamky = ...
zamky.info()

In [ ]:
zamky.head()

### Cvičení

* Spočítej pro každý zámek průměrnou návštěvnost (ze všech dat, která jsou k dispozici).
* Který zámek má nejvyšší a který má nejnižší průměrnou návštěvnost na základě dostupných dat?

In [ ]:
...

Trochu si upravíme data, aby indexem byl rok, a sloupce aby byly názvy objektů. Pomůže nám to pro přehlednost při vykreslování grafu.

In [ ]:
zamky = zamky.transpose()
zamky.index = zamky.index.rename("Rok")

In [ ]:
zamky = zamky.sort_index()
zamky

### Cvičení

* Vykresli graf (např. čarový nebo sloupcový) návštěvnosti jednotlivých zámků v čase.

In [ ]:
...

### Cvičení

* Do nového sloupce `Průměr` spočítej průměrnou návštěvnost zámků za každý rok.
* Vykresli znovu graf i s tímto sloupcem.

In [ ]:
zamky["Průměr"] = ...

In [ ]:
...